src/histogram.rs

```diff
@@ -0,0 +1,365 @@
+// Copyright 2014 The Prometheus Authors
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::convert::From;
+use std::sync::{Arc, RwLock};
+use std::collections::HashMap;
+
+use protobuf::RepeatedField;
+use proto;
+use desc::Desc;
+use errors::{Result, Error};
+use value::make_label_pairs;
+use vec::{MetricVec, MetricVecBuilder};
+use metrics::{Collector, Metric, Opts};
+
+pub const DEFAULT_BUCKETS: &'static [f64; 11] = &[0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0,
+                                                  2.5, 5.0, 10.0];
+
+// `BUCKET_LABEL` is used for the label that defines the upper bound of a
+// bucket of a histogram ("le" -> "less or equal").
+pub const BUCKET_LABEL: &'static str = "le";
+
+#[inline]
+fn check_bucket_lable(label: &str) -> Result<()> {
+    if label == BUCKET_LABEL {
+        return Err(Error::Msg("`le` is not allowed as label name in histograms".to_owned()));
+    }
+
+    Ok(())
+}
+
+/// `HistogramOpts` bundles the options for creating a Histogram metric. It is
+/// mandatory to set Name and Help to a non-empty string. All other fields are
+/// optional and can safely be left at their zero value.
+pub struct HistogramOpts {
+    pub common_opts: Opts,
+
+    // buckets defines the buckets into which observations are counted. Each
+    // element in the slice is the upper inclusive bound of a bucket. The
+    // values must be sorted in strictly increasing order. There is no need
+    // to add a highest bucket with +Inf bound, it will be added
+    // implicitly. The default value is DefBuckets.
+    pub buckets: Vec<f64>,
+}
+
+impl HistogramOpts {
+    pub fn new<S: Into<String>>(name: S, help: S) -> HistogramOpts {
+        HistogramOpts {
+            common_opts: Opts::new(name, help),
+            buckets: Vec::from(DEFAULT_BUCKETS as &'static [f64]),
+        }
+    }
+
+    pub fn namespace<S: Into<String>>(mut self, namesapce: S) -> Self {
```

Comment:
seem duplicated with Opts, do you think a better way to handle all metric options? 
If not, let is keep it here now.